packages/common/src/store/buy-crypto/sagas.ts

```diff
@@ -633,36 +648,230 @@ function* doBuyCryptoViaSol({
+      make({
+        eventName: Name.BUY_CRYPTO_ON_RAMP_FAILURE,
+        mint,
+        provider,
+        amount,
+        error: error.message
+      })
+    )
+  }
+}
+
+/**
+ * Failures happen when the on ramp purchase could not be confirmed, or the
+ * subsequent swap attempts all fail. Get the SOL balance of the wallet, and
+ * if it has a balance greater than min rent, attempt to swap all of it to the
+ * output token. If we get enough of the output token after the swap, it's a
+ * successful recovery.
+ */
+function* recoverBuyCryptoViaSolIfNecessary() {
+  yield* call(waitForAccount)
+  // Pull from context
+  const audiusBackendInstance = yield* getContext('audiusBackendInstance')
+  const getFeatureEnabled = yield* getContext('getFeatureEnabled')
+  const { track, make } = yield* getContext('analytics')
+  const reportToSentry = yield* getContext('reportToSentry')
+
+  // Return early if feature not enabled
+  if (!getFeatureEnabled(FeatureFlags.BUY_USDC_VIA_SOL)) {
+    return
+  }
+
+  // Check for local storage
+  const audiusLocalStorage = yield* getContext('localStorage')
+  const localStorageState = yield* call(
+    [
+      audiusLocalStorage,
+      audiusLocalStorage.getJSONValue<BuyCryptoViaSolLocalStorageState>
+    ],
+    BUY_CRYPTO_VIA_SOL_STATE_KEY
+  )
+
+  // Do nothing if there's no local storage state
+  if (!localStorageState) {
+    return
+  }
+
+  const { mint, amount, provider } = localStorageState
+  yield* call(
+    track,
+    make({
+      eventName: Name.BUY_CRYPTO_RECOVERY_STARTED,
+      mint,
+      provider,
+      amount
+    })
+  )
+
+  // Get config
+  const wallet = yield* call(getRootSolanaAccount, audiusBackendInstance)
+  const connection = yield* call(getSolanaConnection, audiusBackendInstance)
+  const feePayerAddress = yield* waitForValue(getFeePayer)
+  const config = yield* call(getBuyCryptoRemoteConfig, mint)
+  const outputToken = TOKEN_LISTING_MAP[mint.toUpperCase()]
+  let userbank: PublicKey | null = null
+
+  try {
+    if (!feePayerAddress) {
+      throw new BuyCryptoError(
+        BuyCryptoErrorCode.BAD_FEE_PAYER,
+        'Fee Payer is missing'
+      )
+    }
+
+    const outputTokenLamports = 10 ** outputToken.decimals
+    const expectedAmount = Math.ceil(amount * outputTokenLamports)
+    const feePayer = new PublicKey(feePayerAddress)
+    userbank = yield* call(createUserBankIfNeeded, audiusBackendInstance, {
+      mint: localStorageState.mint,
+      feePayerOverride: feePayerAddress,
+      recordAnalytics: track
+    })
+
+    // Get swappable salvage amount
+    const minRent = yield* call(
+      [connection, connection.getMinimumBalanceForRentExemption],
+      0
+    )
+    const balance = yield* call(
+      [connection, connection.getBalance],
+      wallet.publicKey
+    )
+    const salvageInputAmount = balance - minRent
+
+    // Don't do anything if we don't have any SOL.
+    // Don't clear local storage either - maybe the SOL hasn't gotten to us yet?
```

Comment:
I get this use case, but am concerned that it means this state could stay around permanently if the SOL never arrives or was recovered elsewhere etc. I wonder if it's worth the hassle to set expiration conditions on this? Maybe store slot information with the original SOL purchase and detect if we're well beyond that? Or just give it a cutoff of some number of hours or days?

Reply:
Good call